google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorkerPool.java

```diff
@@ -234,47 +233,52 @@ public ApiFuture<AppendRowsResponse> append(StreamWriter streamWriter, ProtoRows
   public ApiFuture<AppendRowsResponse> append(
       StreamWriter streamWriter, ProtoRows rows, long offset) {
     // We are in multiplexing mode after entering the following logic.
-    ConnectionWorker connectionWorker =
-        streamWriterToConnection.compute(
-            streamWriter,
-            (key, existingStream) -> {
-              // Though compute on concurrent map is atomic, we still do explicit locking as we
-              // may have concurrent close(...) triggered.
-              lock.lock();
-              try {
-                // Stick to the existing stream if it's not overwhelmed.
-                if (existingStream != null && !existingStream.getLoad().isOverwhelmed()) {
-                  return existingStream;
-                }
-                // Try to create or find another existing stream to reuse.
-                ConnectionWorker createdOrExistingConnection = null;
-                try {
-                  createdOrExistingConnection =
-                      createOrReuseConnectionWorker(streamWriter, existingStream);
-                } catch (IOException e) {
-                  throw new IllegalStateException(e);
-                }
-                // Update connection to write stream relationship.
-                connectionToWriteStream.computeIfAbsent(
-                    createdOrExistingConnection, (ConnectionWorker k) -> new HashSet<>());
-                connectionToWriteStream.get(createdOrExistingConnection).add(streamWriter);
-                return createdOrExistingConnection;
-              } finally {
-                lock.unlock();
-              }
-            });
+    ConnectionWorker currentConnection;
+    // TODO: Do we need a global lock here? Or is it enough to just lock the StreamWriter?
+    lock.lock();
+    try {
+      currentConnection = streamWriter.getCurrentConnectionPoolConnection();
+      // TODO: Experiment with only checking isOverwhelmed less often per StreamWriter (once per
+      // second?) instead of on every append call.
+      if (currentConnection == null || currentConnection.getLoad().isOverwhelmed()) {
+        // Try to create or find another existing stream to reuse.
+        ConnectionWorker createdOrExistingConnection = null;
+        try {
+          createdOrExistingConnection =
+              createOrReuseConnectionWorker(streamWriter, currentConnection);
+        } catch (IOException e) {
+          throw new IllegalStateException(e);
+        }
+        currentConnection = createdOrExistingConnection;
+        streamWriter.setCurrentConnectionPoolConnection(currentConnection);
+        // Update connection to write stream relationship.
+        // TODO: What if we simply kept an atomic refcount in ConnectionWorker? We could also
+        // manage the refcount in the callback below to precisely track which connections are being
+        // used.
+        currentConnection.getCurrentStreamWriters().add(streamWriter);
```

Comment:
Is this lock good enough to protect currentConnection?

Reply:
we're using it to protect currentConnection.getCurrentStreamWriters()

In theory we could put a lock inside of currentConnection, which would create more granular locking. However this would also cause a lot more lock/unlock activity (e.g. every call to append would have to lock at least two locks) so this change would need measurement to see if it better better.